src/main/java/com/hotels/bdp/circustrain/bigquery/copier/BigQueryCopierFactory.java

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (C) 2018 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.bigquery.copier;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.context.annotation.Profile;
+import org.springframework.core.Ordered;
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+import com.hotels.bdp.circustrain.api.Modules;
+import com.hotels.bdp.circustrain.api.copier.Copier;
+import com.hotels.bdp.circustrain.api.copier.CopierFactory;
+import com.hotels.bdp.circustrain.bigquery.context.CircusTrainBigQueryCondition;
+import com.hotels.bdp.circustrain.bigquery.extraction.BigQueryDataExtractionManager;
+
+@Component
+@Profile({ Modules.REPLICATION })
+@Order(Ordered.HIGHEST_PRECEDENCE)
+@Conditional(CircusTrainBigQueryCondition.class)
+public class BigQueryCopierFactory implements CopierFactory {
+
+  private static final Logger log = LoggerFactory.getLogger(BigQueryCopierFactory.class);
+
+  private final List<CopierFactory> delegates;
+  private final BigQueryDataExtractionManager dataExtractionManager;
+
+  private CopierFactory supportedFactory;
+
+  @Autowired
+  BigQueryCopierFactory(List<CopierFactory> delegates, BigQueryDataExtractionManager dataExtractionManager) {
+    this.delegates = ImmutableList.copyOf(delegates);
+    this.dataExtractionManager = dataExtractionManager;
+  }
+
+  @Override
+  public boolean supportsSchemes(String sourceScheme, String replicaScheme) {
+    for (CopierFactory factory : delegates) {
+      if (factory.supportsSchemes(sourceScheme, replicaScheme)) {
+        supportedFactory = factory;
+        return true;
+      }
```

Comment:
Spent some time exploring implementing a CompositeCopierFactory that filters on CopierFactories that are supported as discussed in follow up meeting. Ran into a couple of issues for this use case, we have CopierFactories that return true by default (DistCPCopierFactory) and if we chain the copies any copier on the classpath that by default returns true in its supportsSchemes method will execute which isn't desirable, current implementation executes one and only one copy. Tried to work around this but its quite ugly, and we already have Spring wired up nicely to execute one and only one copy job so seemed a bit weird to do it explicitly.. Thoughts? Will push code in another branch

Reply:
Not sure how much better that approach is. If we do decide to stick with this then at the very least we should put some javadoc/comments on the newInstance() method and/or this method to indicate that it relies on CT calling these things in order for this to actually work.

Reply:
I prefer this PR. Yeah the whole ordered list of copierFactories that you need to pick one from seems to work against us a bit.
Thinking out loud, would it simplify things if we don't rely on this `List<CopierFactory> delegates` but just explicitly reference the Factories we support. We'll get multiple BigQueryCopierFactories, e.g. 
```
BigQueryToS3CopierFactory(S3MapreduceCopierFactory factory) {
      supportsScheme() {
        sourceScheme.startsWith("gs") && targetScheme.startsWith("s3")
     }
    }
```
And another `BigQueryToHDFSCopierFactory` which takes the appropriate factory.

That way we'll stop relying on ct so much??? Thoughts?

Reply:
I don't think we should circumvent the typical Spring plugin behavior, that is defined in Circus Train, although this is a plugin the ability for users to also define their own plugins like Copiers is a nice to have. If we start hardwiring copiers, it prevents hypothetical users from designing their own, and also leaves opportunities for bugs if we change core CT copiers and don't update this project accordingly

Reply:
@patduin thoughts on the above?

Reply:
I'm fine with what we got